### PR TITLE
Sheet: Control animation behavior with AnimationContext

### PR DESCRIPTION
**Pull request by themindfuldev**
This PR extracts the animation handling logic from `<Sheet>` into `<AnimationController>` while providing a number of new options:
- A clear API to separate callbacks for `onDismissStart` (when a dismiss animation begins) and `onDismissEnd` (when a dismiss animation is over and only then unmount/cleanup will take place)
- A Context Provider `AnimationController` and a custom hook `useAnimation` to make it easier to control this animation behavior on internal components.
- 4 distinct animation states: in -> postIn -> out -> postOut
- A plug-and-play configuration where all the behavior works fine when an `<AnimationController>` is not provided
- Respecting `useReducedMotion` for a11y

## Test Plan
- Tested on `<Sheet>` docs
- Tested with OS reduced motion activated

### CODE REVIEW

File: packages/gestalt/src/AnimationController.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    [setAnimationState, shouldAnimate]
+  );
+
+  useEffect(() => {
+    if (animationState === 'post-out') {
+      onDismissEnd();
+    }
+  }, [animationState, onDismissEnd]);
+
+  return animationState !== 'post-out' ? (
```
Comment (dangerismycat) [inline packages/gestalt/src/AnimationController.js:60]: For readability, consider phrasing this as an early return:
```
if (animationState === 'post-out') {
  return null;
}

return ( ...
```

File: packages/gestalt/src/AnimationController.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    }
+  }, [animationState, setAnimationState]);
+  return {
+    animationState,
+    onAnimationEnd: animationState ? onAnimationEnd : null,
+  };
+}
+
+export default {
+  AnimationController,
```
Comment (dangerismycat) [inline packages/gestalt/src/AnimationController.js:86]: Should the controller itself be the default return?

**Review by dangerismycat — APPROVED** [review]
This is great!

File: packages/gestalt/src/AnimationController.js (hunk 2 above)
Comment (themindfuldev) [reply packages/gestalt/src/AnimationController.js:86]: It makes sense, I agree :)

**Comment by christianvuerings** [comment]
Checked in with @themindfuldev offline - we will be importing the animation controller inside of `Sheet`

File: docs/src/AnimationController.doc.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+const cards: Array<Node> = [];
+const card = c => cards.push(c);
+
+card(
+  <PageHeader
+    name="AnimationController"
+    description={`An \`<AnimationController>\` is a wrapper to control animation of its wrapped components. 
+    It works with a \`Context.Provider\` which holds the state for \`animationState\`. 
+    Finally, it provides the custom hook \`useAnimation\` which provides an object \`{ animationState, onAnimationEnd }\` where:
+    - \`animationState\`: the current animation state. Possible values are: \`null\`, \`"in"\`, \`"post-in"\`, \`"out"\`, \`"post-out"\`.
```
[the remark this answers is not in the document]
Comment (themindfuldev) [reply docs/src/AnimationController.doc.js]: Changed :)

**Comment by themindfuldev** [comment]
Updated docs: https://deploy-preview-1236--gestalt.netlify.app/Sheet

File: docs/src/Sheet.doc.js [hunk 4]
```diff
@@ -307,27 +377,27 @@ function SheetRefExample() {
...
               accessibilitySheetLabel="Focused sheet"
-              onDismiss={() => { setShowSheet(!showSheet) }}
+              onDismiss={() => setShouldShow(false)}
               ref={sheetRef}
               size="md"
             >
               <Box color="white" minHeight={400} padding={8}>
                 <Box marginBottom={4}>
-                  <Heading size="md">Focused content</Heading>
+                  <Heading size="md">Focused content</Heading>
```
Comment (christianvuerings) [inline docs/src/Sheet.doc.js:394]: @themindfuldev looks like it's adding a bunch of extra whitespace after certain lines, can we remove that?
image

**Review by christianvuerings — APPROVED** [review]

File: docs/src/Sheet.doc.js (hunk 4 above)
Comment (themindfuldev) [reply docs/src/Sheet.doc.js:394]: Sure - removing as part of https://github.com/pinterest/gestalt/pull/1242

**Pull request merged**